indexing-service/src/main/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactory.java

```diff
@@ -286,6 +209,65 @@ public WindowedStorageAdapter apply(final PartitionChunk<DataSegment> input)
     catch (IOException | SegmentLoadingException e) {
       throw Throwables.propagate(e);
     }
+  }
+
+  @VisibleForTesting
+  static List<String> getUniqueDimensions(
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      InputRowParser inputRowParser
+  )
+  {
+    final Set<String> exclusions = inputRowParser.getParseSpec().getDimensionsSpec().getDimensionExclusions();
+    final BiMap<String, Integer> uniqueDims = HashBiMap.create();
+
+    // Here, we try to retain the order of dimensions as they were specified since the order of dimensions may be
+    // optimized for performance.
+    // Dimensions are extracted from the recent segments to olders because recent segments are likely to be queried more
+    // frequently, and thus the performance should be optimized for recent ones rather than old ones.
+
+    // timelineSegments are sorted in order of interval
+    int index = 0;
+    for (TimelineObjectHolder<String, DataSegment> timelineHolder : Lists.reverse(timelineSegments)) {
+      for (PartitionChunk<DataSegment> chunk : timelineHolder.getObject()) {
+        for (String dimension : chunk.getObject().getDimensions()) {
+          if (!uniqueDims.containsKey(dimension) && !exclusions.contains(dimension)) {
+            uniqueDims.put(dimension, index++);
+          }
+        }
+      }
+    }
+
+    final BiMap<Integer, String> orderedDims = uniqueDims.inverse();
+    return IntStream.range(0, orderedDims.size())
+                    .mapToObj(orderedDims::get)
+                    .collect(Collectors.toList());
+  }
+
+  @VisibleForTesting
+  static List<String> getUniqueMetrics(List<TimelineObjectHolder<String, DataSegment>> timelineSegments)
+  {
+    final BiMap<String, Integer> uniqueMetrics = HashBiMap.create();
+
+    // Here, we try to retain the order of metrics as they were specified since the order of metrics may be
```

Comment:
Unlike dimensions, order of metrics doesn't matter for performance. Dimension order matters because it affects sorting of the rows and can be used to improve locality (rows are sorted by time first, but rows within the same time bucket are sorted by dimensions, in order). But metric order doesn't affect sorting.

Reply:
Right. Metrics currently don't affect to performance. However, we're going to handle dimensions and metrics in the same way. So, I guess this will be needed in the future. Do you think it's better to add later?

BTW, I updated the comments here.